coordinator/src/main/java/org/apache/uniffle/coordinator/AppBalanceSelectStorageStrategy.java

```diff
@@ -39,106 +45,127 @@
+            fos.write(data);
+            fos.flush();
+          }
+          byte[] readData = new byte[fileSize];
+          int readBytes;
+          try (FSDataInputStream fis = fs.open(testPath)) {
+            int hasReadBytes = 0;
+            do {
+              readBytes = fis.read(readData);
+              if (hasReadBytes < fileSize) {
+                for (int i = 0; i < readBytes; i++) {
+                  if (data[hasReadBytes + i] != readData[i]) {
+                    RankValue rankValue = remoteStoragePathRankValue.get(uri);
+                    remoteStoragePathRankValue.put(uri,
+                        new RankValue(Long.MAX_VALUE, rankValue.getAppNum().get()));
+                    throw new RssException("The content of reading and writing is inconsistent.");
+                  }
+                }
+              }
+              hasReadBytes += readBytes;
+            } while (readBytes != -1);
+          }
         }
-      } else {
-        LOG.warn("Can't find counter for remote storage: {}", storagePath);
-        remoteStoragePathCounter.putIfAbsent(storagePath, new RankValue(0));
-      }
-      if (remoteStoragePathCounter.get(storagePath).getAppNum().get() == 0
-          && !availableRemoteStorageInfo.containsKey(storagePath)) {
-        remoteStoragePathCounter.remove(storagePath);
+      } catch (Exception e) {
+        setRemotePathIsHealthy(false);
+        LOG.error("Storage read and write error, we will not use this remote path {}.", uri, e);
+        RankValue rankValue = remoteStoragePathRankValue.get(uri);
+        remoteStoragePathRankValue.put(uri, new RankValue(Long.MAX_VALUE, rankValue.getAppNum().get()));
+      } finally {
+        return sortPathByRankValue(uri, rssTest, remotePathIsHealthy);
       }
+    } else {
+      return Lists.newCopyOnWriteArrayList(remoteStoragePathRankValue.entrySet());
     }
   }
 
+  /**
+   * When choosing the AppBalance strategy, each time you select a path,
+   * you should know the number of the latest apps in different paths
+   */
   @Override
-  public synchronized void removePathFromCounter(String storagePath) {
-    RankValue atomic = remoteStoragePathCounter.get(storagePath);
-    if (atomic != null && atomic.getAppNum().get() == 0) {
-      remoteStoragePathCounter.remove(storagePath);
+  public synchronized RemoteStorageInfo pickStorage(
+      List<Map.Entry<String, RankValue>> uris, String appId) {
+    boolean isUnhealthy =
+        uris.stream().noneMatch(rv -> rv.getValue().getReadAndWriteTime().get() != Long.MAX_VALUE);
+    if (!isUnhealthy) {
+      // If there is only one unhealthy path, then filter that path
+      uris = uris.stream().filter(rv -> rv.getValue().getReadAndWriteTime().get() != Long.MAX_VALUE).sorted(
+          Comparator.comparingInt(entry -> entry.getValue().getAppNum().get())).collect(Collectors.toList());
+    } else {
+      // If all paths are unhealthy, assign paths according to the number of apps
```

Comment:
Should we add some logs and metrics when all paths are unhealthy?

Reply:
OK, I will add.